Implement custom tester functionality in Skaffold

PriyaModali opened the pull request:
Fixes: ##5395

**Description**
Add new custom tester functionality to Skaffold. This feature enable users to run custom tests in the Skaffold pipeline.


The test command:
1. Implements Runner interface
2. Enables ability to run custom commands between build & deploy phases

Run ```skaffold dev``` to run the command in the Skaffold pipeline.

Added unit tests and integrations tests for the newly added test command.

**Related**:  
#5333

Note:
Right now, changing a test dependency file does not trigger a build but triggers only deployment.

Follow up work:
Changing a test dependency file does not trigger a build but will trigger test.


### Testing notes

> `skaffold dev` - first loop what does the output in test phase
```
Testing images...
Running custom test command: "./test.sh" with timeout 60 s
go custom test 
ok      github.com/GoogleContainerTools/skaffold/integration/examples/custom-tests      0.812s
Command finished successfully
Running custom test command: "echo Hello world!!"
Hello world!!
Command finished successfully
Tags used in deployment:
 - skaffold-example -> skaffold-example:0a3c7c9f048a204d72b2e4dbc017351754ad8d283505d9cd43e54023eb073ee5
Starting deploy...
 - pod/getting-started created
Waiting for deployments to stabilize...
Deployments stabilized in 42.507007ms
Press Ctrl+C to exit
Watching for changes...
[getting-started] ...
....
```

> `skaffold dev` - changing a test file does not trigger a build
```
Testing images...
Running custom test command: "./test.sh" with timeout 60 s
go custom test 
ok      github.com/GoogleContainerTools/skaffold/integration/examples/custom-tests      0.812s
Command finished successfully
Running custom test command: "echo Hello world!!"
Hello world!!
Command finished successfully
Tags used in deployment:
 - skaffold-example -> skaffold-example:5b2a46985b6af5b3a545443d67ae6198c100080d897ffd900db702eca7eea9ba
Starting deploy...
 - pod/getting-started created
Waiting for deployments to stabilize...
Deployments stabilized in 43.06159ms
Press Ctrl+C to exit
Watching for changes...
[getting-started] ...
....
....
Tags used in deployment:
 - skaffold-example -> skaffold-example:5b2a46985b6af5b3a545443d67ae6198c100080d897ffd900db702eca7eea9ba
Starting deploy...
Waiting for deployments to stabilize...
Deployments stabilized in 54.700905ms
Watching for changes...
[getting-started] ...
....
....
```

> `skaffold dev` - changing build dependency triggers build and test
```
Testing images...
Running custom test command: "./test.sh" with timeout 60 s
go custom test 
ok      github.com/GoogleContainerTools/skaffold/integration/examples/custom-tests      0.812s
Command finished successfully
Running custom test command: "echo Hello world!!"
Hello world!!
Command finished successfully
Tags used in deployment:
 - skaffold-example -> skaffold-example:bfe229ee3daf1af14d4db031daf84ebdcb9ddd97a0530fa8fa8d52e66d4c0780
Starting deploy...
 - pod/getting-started created
Waiting for deployments to stabilize...
Deployments stabilized in 49.206668ms
Press Ctrl+C to exit
Watching for changes...
[getting-started] ...
....
....
Generating tags...
 - skaffold-example -> skaffold-example:v1.19.0-85-g55659bd41-dirty
Found [minikube] context, using local docker daemon.
Building [skaffold-example]...
Sending build context to Docker daemon  3.584kB
....
....
Successfully built d21f428215c1
Successfully tagged skaffold-example:v1.19.0-85-g55659bd41-dirty
Testing images...
Running custom test command: "./test.sh" with timeout 60 s
go custom test 
ok      github.com/GoogleContainerTools/skaffold/integration/examples/custom-tests      0.812s
Command finished successfully
Running custom test command: "echo Hello world!!"
Hello world!!
Command finished successfully
Tags used in deployment:
 - skaffold-example -> skaffold-example:d21f428215c1ac7896825508279f6dd06830c93d5a8f189baf8627c58e5d353f
Starting deploy...
 - pod/getting-started configured
Waiting for deployments to stabilize...
Deployments stabilized in 24.410772ms
Watching for changes...
[getting-started] ...
....
....
```

> `skaffold dev` - Test failure with a timeout
```
Testing images...
Running custom test command: "./test.sh" with timeout 60 s
go custom test 
ok      github.com/GoogleContainerTools/skaffold/integration/examples/custom-tests      0.812s
Command timed out
Pruning images...
exiting dev mode because test failed after first build: running tests: running custom test command: context deadline exceeded
```

> `skaffold dev` - Test failure with a non-0 exit code
```
Testing images...
Running custom test command: "./test.sh" with timeout 60 s
go custom test 
ok      github.com/GoogleContainerTools/skaffold/integration/examples/custom-tests      0.812s
Pruning images...
exiting dev mode because test failed after first build: running tests: running custom test command: Command finished with non-0 exit code: exit status 20
```

briandealwis commented on pkg/skaffold/test/custom/custom.go:
```diff
@@ -0,0 +1,161 @@
+}
+
+// TestDependencies returns dependencies listed for a custom test
+func (ct *Runner) TestDependencies() ([]string, error) {
```
To disambiguate this from a normal Test* function, maybe consider naming this `ListTestDependencies()`?

PriyaModali replied:
This is an interface method that's being called from New() in dev loop.

briandealwis reviewed (commented):
Here are some quick comments

PriyaModali commented on pkg/skaffold/test/custom/custom.go:
```diff
@@ -0,0 +1,161 @@
+	color.Default.Fprintln(out, "Custom test timeout is set to: ", test.TimeoutSeconds)
+
+	if test.TimeoutSeconds > 0 {
+		newCtx, cancel := context.WithTimeout(ctx, (time.Duration(test.TimeoutSeconds))*(time.Second))
```
Removed.

PriyaModali commented on pkg/skaffold/schema/latest/config.go:
```diff
@@ -1012,6 +1015,34 @@ type CustomDependencies struct {
+}
+
+// CustomTestDependencies is used to specify dependencies for custom test command.
+// `paths` should be specified for file watching to work as expected.
```
I guess we try to read it & fail?

PriyaModali commented on examples/custom-tests/skaffold.yaml:
```diff
@@ -0,0 +1,17 @@
+test:
+  - image: skaffold-example
+    custom:
+      - command: ./command/test.sh #go run example.go 5 10
```
The script sample shows the timer functionality.

PriyaModali commented on examples/custom-tests/main.go:
```diff
@@ -0,0 +1,14 @@
+	"time"
+)
+
+func main() {
```
This `main()` is for the dev loop. The one in `command/` is fo runnig the test.

PriyaModali commented on examples/custom-tests/command/test.sh:
```diff
@@ -0,0 +1,35 @@
+    go run ./command/basic.go
+
+    i=$(($i + 5))   
+    sleep 5
```
It runs every 5 seconds in a loop to show the timer functionality.

PriyaModali commented on examples/custom-tests/command/test.sh:
```diff
@@ -0,0 +1,35 @@
+i=0
+while true 
+do
+    go clean 
```
Removed.

PriyaModali commented on examples/custom-tests/command/test.sh:
```diff
@@ -0,0 +1,35 @@
+# This script runs go test with a better output:
+# - It prints the failures in RED
+# - It recaps the failures at the end
+# - It lists the 20 slowest tests
```
Removed. The script doesn't do this now.

PriyaModali commented on examples/custom-tests/README.md:
```diff
@@ -0,0 +1,25 @@
+test:
+    - image: skaffold-example
+    Custom:
+        - command: <command>
```
Added dependencies.

PriyaModali commented on pkg/skaffold/test/custom/custom.go:
```diff
@@ -0,0 +1,161 @@
+
+	var cmd *exec.Cmd
+	// We evaluate the command with a shell so that it can contain
+	// env variables.
```
Updated.

PriyaModali commented on pkg/skaffold/test/custom/custom.go:
```diff
@@ -0,0 +1,161 @@
+}
+
+// New creates a new custom.Runner.
+func New(cfg docker.Config, wd string, ct latest.CustomTest) (*Runner, error) {
```
The naming is in sync with the existing structure tests:
https://github.com/GoogleContainerTools/skaffold/blob/27d1de28da11a8676c6f9913affce066e807083f/pkg/skaffold/test/structure/structure.go#L43

tejal29 replied:
I had given the feedback previously to avoid stuttering to replace `test/custom.NewCusotmTestRunner`  with just `test/custom.New`

briandealwis commented on integration/examples/custom-tests/command/basic_test.go:
```diff
@@ -0,0 +1,51 @@
+limitations under the License.
+*/
+
+// Note: This is an example from `Go by Example` site (https://gobyexample.com/testing)
```
This code is licensed by CC-BY 3.0; I think this is ok to include but we maybe should check (https://opensource.google/docs/thirdparty/licenses/) — it's unusual to have Creative Commons on code.  To be honest, I think you should create a new example from scratch.

PriyaModali replied:
Added these tests per the suggestion here - https://github.com/GoogleContainerTools/skaffold/pull/5394#discussion_r575510745

I'll revisit the sample.

PriyaModali replied:
Updated the sample.

briandealwis commented on pkg/skaffold/test/custom/custom_test.go:
```diff
@@ -0,0 +1,71 @@
+					Command:        "echo Running Custom Test command1.",
+					TimeoutSeconds: 10,
+					Dependencies: &latest.CustomTestDependencies{
+						Command: "echo [\"file1\",\"file2\",\"file3\"]",
+						Paths:   []string{"**"},
+						Ignore:  []string{"b*"},
```
The implementation seems to use either command OR paths+ignore, but not both.  Can you add a validation check in pkg/skaffold/schema/validation?

PriyaModali replied:
Added schema validation for custom test fields.

PriyaModali commented on integration/examples/custom-tests/README.md:
````diff
@@ -0,0 +1,33 @@
+          paths: <file dependencies>
+          - <paths glob>
+```
+
````
Yes.

PriyaModali commented on pkg/skaffold/test/custom/custom_test.go:
```diff
@@ -0,0 +1,71 @@
+			tests: []*latest.TestCase{{
+				ImageName:      "image",
+				StructureTests: []string{"test.yaml"},
+				CustomTests: []latest.CustomTest{{
+					Command:        "echo Running Custom Test command1.",
+					TimeoutSeconds: 10,
+					Dependencies: &latest.CustomTestDependencies{
+						Command: "echo [\"file1\",\"file2\",\"file3\"]",
+						Paths:   []string{"**"},
+						Ignore:  []string{"b*"},
+					},
+				}},
+			}},
```
Thanks for the suggestion. Updates the test.

PriyaModali commented on pkg/skaffold/test/custom/custom.go:
```diff
@@ -0,0 +1,160 @@
+
+	switch {
+	case test.Dependencies.Command != "":
+		split := strings.Split(test.Dependencies.Command, " ")
```
The implementation is line with the custom artifact dependencies https://github.com/GoogleContainerTools/skaffold/blob/827139dccb7eea76578db30bec40376ddc9de689/pkg/skaffold/build/custom/dependencies.go#L38

tejal29 replied:
agree with @briandealwis. Makes sense to keep it consistent with Test command run by executing it with "sh -c" or "cmd.exe /c"

PriyaModali replied:
Made it consistent with Custom Test command.

PriyaModali commented on pkg/skaffold/test/custom/custom.go:
```diff
@@ -0,0 +1,160 @@
+type Runner struct {
+	customTest     latest.CustomTest
+	testWorkingDir string
+	localDaemon    docker.LocalDaemon
```
It is currently used for querying the environment & will be used to pull the image down for the custom command.

tejal29 replied:
I don't think we need the localDaemon and the environment vairables from local daemon for custom tests.

PriyaModali replied:
Removed localDaemon

PriyaModali commented on integration/examples/custom-tests/skaffold.yaml:
```diff
@@ -0,0 +1,17 @@
+test:
+  - image: skaffold-example
+    custom:
+      - command: ./command/test.sh
```
Updated the sample to run a simple script with echo command.

PriyaModali commented on pkg/skaffold/test/custom/custom_test.go:
```diff
@@ -0,0 +1,71 @@
+		}
+
+		testRunner, err := New(cfg, cfg.workingDir, custom)
+		t.CheckNoError(err)
```
Added tests for error cases.

briandealwis commented on pkg/skaffold/test/custom/custom.go:
```diff
@@ -0,0 +1,160 @@
+	cmd.Stderr = out
+	cmd.Env = ct.env()
+
+	if err := cmd.Start(); err != nil {
+		return fmt.Errorf("error running cmd: %w, output: %v", err, out)
+	}
+
+	if err := cmd.Wait(); err != nil {
```
Should be able to just use:
```suggestion
	if err := util.RunCmd(cmd); err != nil {
```

PriyaModali commented on integration/examples/custom-tests/main.go:
```diff
@@ -0,0 +1,38 @@
+	rand.Seed(time.Now().UnixNano())
+	x := rand.Intn(100)
+
+	rand.Seed(time.Now().UnixNano())
```
Removed reseeding. Missed in the previous review.

PriyaModali commented on integration/examples/custom-tests/main.go:
```diff
@@ -0,0 +1,38 @@
+
+func main() {
+	rand.Seed(time.Now().UnixNano())
+	x := rand.Intn(100)
+
+	rand.Seed(time.Now().UnixNano())
+	y := rand.Intn(100)
+
+	min := MinInt(x, y)
+	fmt.Println("Min of ", x, " and ", y, " is: ", min)
```
Made it a loop.

PriyaModali commented on integration/examples/custom-tests/README.md:
```diff
@@ -0,0 +1,32 @@
+    - image: skaffold-example
+      custom:
+      - command: <command>
+        timeout: <timeout in seconds>
```
renamed config key name to timeoutSeconds

PriyaModali commented on integration/examples/custom-tests/main.go:
```diff
@@ -0,0 +1,14 @@
+package main
```
I renamed basic.go to main.go

tejal29 replied:
Sorry, this could have been from the reviewing i started Friday last week.

tejal29 commented on integration/examples/custom-tests/skaffold.yaml:
```diff
@@ -7,7 +7,7 @@ test:
     custom:
       - command: ./test.sh
-        timeout: 60 #in seconds
+        timeoutSeconds: 60 #in seconds
```
you can get rid of "#in seconds" here.

tejal29 commented on pkg/skaffold/schema/validation/validation.go:
```diff
@@ -479,3 +480,27 @@ func validateSingleKubeContext(configs []*latest.SkaffoldConfig) []error {
+// validateCustomTest
+// - makes sure that command is not empty
+// - makes sure that dependencies.ignore is only used in conjunction with dependencies.paths
+func validateCustomTest(tcs []*latest.TestCase) (errs []error) {
```
Can we add tests for this function in `validation_test.go`

PriyaModali replied:
Added tests.

tejal29 commented on pkg/skaffold/test/custom/custom.go:
```diff
@@ -0,0 +1,156 @@
+}
+
+// env returns a merged environment of the current process environment and any extra environment.
+func (ct *Runner) env() []string {
```
This is not needed for custom tests. 
container-structure-tests relies on localDaemon to pull image as it cannot pull imgaes from remote.

PriyaModali replied:
Removed localDaemon for now.

tejal29 replied:
We want to just send the image name to the custom command & don't need to pull it down?  - that is correct.

tejal29 commented on pkg/skaffold/test/custom/custom.go:
```diff
@@ -0,0 +1,156 @@
+			select {
+			case <-ctx.Done():
+				if ctx.Err() == context.DeadlineExceeded {
+					fmt.Println("Command timed out")
```
This should go to `out` using colored output writer similar to all other messages. 
Also makes sense to add a distinct error code instead of returning ctx.Err

PriyaModali replied:
Added colored output. 
Distinct error codes will be added in a follow up pr #5431.

tejal29 commented on pkg/skaffold/test/custom/custom.go:
```diff
@@ -0,0 +1,156 @@
+	cmd.Stderr = out
+	cmd.Env = ct.env()
+
+	if err := util.RunCmd(cmd); err != nil {
```
Should we use `util.RunCmdOut` to capture the test logs and make it user friendly?
1) stream test script logs as they become avaliable

tejal29 replied:
I am also thinking if we shd switch to using `exec.Start` and graceful termination on cancelling like 
https://github.com/GoogleContainerTools/skaffold/blob/dafb4152be74c7cd7d363790f2099f2b6f057ac3/pkg/skaffold/build/custom/script.go#L41

PriyaModali replied:
I considered graceful termination but current implementation of graceful termination doesn't accommodate timer.

tejal29 replied:
hmm so right now, users will only know if a command failed or not? The DD mentions, skaffold will be logging output from the command https://docs.google.com/document/d/1aMfQtlf1gEFbMMVbWy2Iep9WQalLHou6Myc9EVbz01w/edit?ts=6007c302&resourcekey=0-T7xncfS8izrFHTaMeFtMfg#heading=h.7008ql63pa1

We need to support this.

PriyaModali replied:
I was originally using exec.Start() but updated to use exec.Run() per this previous feedback
https://github.com/GoogleContainerTools/skaffold/pull/5451#discussion_r584945883

tejal29 replied:
using `util.RunCmd` is not going to show the command output right?
can you verify this and update the PR description?

PriyaModali replied:
`util.RunCmd` does show the command output.

tejal29 commented on pkg/skaffold/test/custom/custom_test.go:
```diff
@@ -0,0 +1,119 @@
+	"github.com/GoogleContainerTools/skaffold/testutil"
+)
+
+func TestNewCustomTestRunner(t *testing.T) {
```
Can we add some tests for `TestDependencies`
https://github.com/GoogleContainerTools/skaffold/blob/dafb4152be74c7cd7d363790f2099f2b6f057ac3/pkg/skaffold/build/custom/dependencies_test.go

PriyaModali replied:
Added unit tests for TestDependencies

gsquared94 commented on integration/examples/custom-tests/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+# for `skaffold debug` (https://skaffold.dev/docs/workflows/debug/).
+ENV GOTRACEBACK=single
+CMD ["./app"]
+COPY --from=builder /app .
```
add newline at end of file

gsquared94 commented on pkg/skaffold/schema/latest/config.go:
```diff
@@ -1012,6 +1015,36 @@ type CustomDependencies struct {
+
+	// Paths should be set to the file dependencies for this command, so that the skaffold file watcher knows when to retest and perform file synchronization.
+	// For example: `["src/test/**"]`
+	Paths []string `yaml:"paths,omitempty" yamltags:"oneOf=dependency"`
```
add tag `skaffold:"filepath"` to `Paths` and `Ignore`

PriyaModali replied:
Added.

gsquared94 replied:
can you please verify that your set of changes work with [this](https://github.com/GoogleContainerTools/skaffold/pull/5472) PR (in case your's gets merged earlier)

PriyaModali commented:
@briandealwis & @tejal29, thank you for the reviews. I have addressed all the feedback. Welcome more. Otherwise, I'd like to get this into a good state for approval so that I can move on to #5431.

tejal29 commented:
> @briandealwis & @tejal29, thank you for the reviews. I have addressed all the feedback. Welcome more. Otherwise, I'd like to get this into a good state for approval so that I can move on to #5431.

Thanks @PriyaModali. Can you mention little bit in the description what are the output changes. It be great to include examples on the `skaffold dev` on the `custom-tests` example you have.

PriyaModali commented:
> > @briandealwis & @tejal29, thank you for the reviews. I have addressed all the feedback. Welcome more. Otherwise, I'd like to get this into a good state for approval so that I can move on to #5431.
> 
> Thanks @PriyaModali. Can you mention little bit in the description what are the output changes. It will be great to include example running `skaffold dev` on the `custom-tests` example you have.
> 
> I would probably do the following testing and add Testing notes e.g. [0](https://github.com/GoogleContainerTools/skaffold/pull/5121), [1](https://github.com/GoogleContainerTools/skaffold/pull/4828#issuecomment-705245580)
> 
> 1. skaffold dev, - first loop what does the output in test phase look like
> 2. skaffold dev - changing a test file does not trigger a build but only test
> 3. skaffold dev - changing build dependency triggers build and test.

Added Testing notes to the description.

tejal29 commented on pkg/skaffold/test/custom/custom_test.go:
```diff
@@ -0,0 +1,232 @@
+				workingDir: tmpDir.Root(),
+				tests: []*latest.TestCase{{
+					ImageName:      "image",
+					StructureTests: []string{"test.yaml"},
```
you can remove this `StructureTests` here right? This has no effect on the custom test runner.

tejal29 commented on pkg/skaffold/test/custom/custom_test.go:
```diff
@@ -0,0 +1,232 @@
+			Command:        "echo Running Custom Test command.",
+			TimeoutSeconds: 10,
+			Dependencies: &latest.CustomTestDependencies{
+				Command: "echo [\"file1\",\"file2\",\"file3\"]",
+				Paths:   []string{"**"},
```
specifying dependencies  paths and command is not a valid config right? Can we have valid examples in the test?

tejal29 commented on pkg/skaffold/test/custom/custom_test.go:
```diff
@@ -0,0 +1,232 @@
+		testutil.Run(t, "Testing new custom test runner", func(t *testutil.T) {
+			tmpDir := t.NewTempDir().Touch("test.yaml")
+
+			if runtime.GOOS == Windows {
+				test.expectedError = "exit status"
+			}
```
This is strange and i haven't seen this in test. I think you can get rid of this by overriding the function the `util.RunCmd` and return specific error message or nil error.

PriyaModali replied:
`util.RunCmd` does not check for the platform.

There are many special cases in the tests of skaffold projects for the platform specific errors. Here is an example:
https://github.com/GoogleContainerTools/skaffold/blob/ba2390db18ad15ccf1eefbce39d6dd8f34016914/pkg/skaffold/kubernetes/portforward/kubectl_forwarder_test.go#L148

tejal29 replied:
the point is to not run the command but mock it using overrides to make sure
1) the command is called with right params
2) does the right thing when error or no error is returned.

PriyaModali replied:
Updated tests to use mock by overriding util.RunCmd method.

tejal29 commented on pkg/skaffold/test/custom/custom.go:
```diff
@@ -0,0 +1,141 @@
+		color.Default.Fprintf(out, "Running custom test command: %q\n", command)
+	} else {
+		color.Default.Fprintf(out, "Running custom test command: %q with timeout %d s\n", command, test.TimeoutSeconds)
+		newCtx, cancel := context.WithTimeout(ctx, time.Duration(test.TimeoutSeconds)*time.Second)
```
`time.Duration(test.TimeoutSeconds)*time.Second)` can be simplified as `time.Second * test.TimeoutSeconds`
e.g. https://play.golang.org/p/uP0vtklUJ6r

PriyaModali replied:
Getting the below error with the suggested change:
`invalid operation: time.Second * test.TimeoutSeconds (mismatched types time.Duration and int)`

tejal29 replied:
have you tried matching the types in the playground url I gave?

```
newCtx, cancel := context.WithTimeout(context.Background(), time.Second * 10)

```

PriyaModali replied:
Yes. Updated.

tejal29 replied:
ah. looks like `test.TimeoutSeconds` is int. However, in my example 10 is considered as `int64`

You can change it back to what you had before,

PriyaModali replied:
Reverted the changes.

tejal29 commented on pkg/skaffold/schema/validation/validation_test.go:
```diff
@@ -1276,3 +1276,63 @@ func TestValidateTaggingPolicy(t *testing.T) {
+			},
+			expectedErrors: 1,
+		}, {
+			description: "ignore either path or command",
```
the test description is incorrect.

tejal29 commented on pkg/skaffold/test/custom/custom.go:
```diff
@@ -0,0 +1,141 @@
+		return fmt.Errorf("unable to parse test command %q: %w", test.Command, err)
+	}
+
+	logrus.Infof("Running custom test command %q", command)
```
We can get rid of this `logrus.Infof` as it not adding any extra information but repetition of what is below.

tejal29 commented on pkg/skaffold/test/custom/custom_test.go:
```diff
@@ -0,0 +1,232 @@
+			workingDir: tmpDir.Root(),
+			tests: []*latest.TestCase{{
+				ImageName:      "image",
+				StructureTests: []string{"test.yaml"},
```
please get rid of the structure test definition here and no need to create an empty `test.yaml`

tejal29 commented on pkg/skaffold/test/custom/custom_test.go:
```diff
@@ -0,0 +1,232 @@
+
+		testRunner, err := New(cfg, cfg.workingDir, custom)
+		t.CheckNoError(err)
+		err = testRunner.Test(context.Background(), ioutil.Discard, nil)
```
Please use `test.Override` and not actually rely on executing the command in the unit test.
We want to limit the the scope of unit test to
1) making sure the `util.RunCmd` is called with correct parameters.
2) The function does the right thing based on your mock return value.

PriyaModali replied:
This design is similar to both structure tests & the main test.

tejal29 replied:
I am not talking about the mock config. I am talking about the command in test cases "echo" etc actually being executed. They should not be executed via mocking the `util.RunCmd` method.

PriyaModali replied:
Updated tests to use mock by overriding util.RunCmd method.

tejal29 commented on pkg/skaffold/test/custom/custom_test.go:
```diff
@@ -0,0 +1,232 @@
+			expectedError: "exit status 20",
+		},
+		{
+			description: "Command timed out",
+			custom: latest.CustomTest{
+				Command:        "sleep 20",
+				TimeoutSeconds: 2,
+			},
+			shouldErr:     true,
+			expectedError: "context deadline exceeded",
```
I would suggest removing the sleep in the unit test and this test case for `timeout`. 
Adding sleep in tests is not recommended https://ardalis.com/thread-sleep-in-tests-is-evil/.
If you want to test the context deadline exceeded case, please Use overrides wherever possible.

tejal29 commented on pkg/skaffold/test/custom/custom_test.go:
```diff
@@ -0,0 +1,232 @@
+				"[\"file1\",\"file2\",\"file3\"]",
+			))
+		} else {
+			t.Override(&util.DefaultExecCommand, testutil.CmdRunOut(
```
Use the same apprach for `TestCustomCommandError` and `TestNewCustomTestRunner`

PriyaModali replied:
There are many special cases in the tests of skaffold projects for the platform specific errors. Here is an example:
https://github.com/GoogleContainerTools/skaffold/blob/ba2390db18ad15ccf1eefbce39d6dd8f34016914/pkg/skaffold/kubernetes/portforward/kubectl_forwarder_test.go#L148

tejal29 replied:
please change it.  it is not required in this unit test. The test you pointed actually checking if a command is running in background and has extra logic of terminating the command in unit test.

PriyaModali replied:
If I don't special case it, the test will fail on the PR. This is because the actual commands run are different on different platforms:
On Windows it is `cmd.exe /C`
On non-windows it is `sh -c`

See the code below:
https://github.com/GoogleContainerTools/skaffold/blob/55659bd416070ead13e9a3fe65519ec57cdceb9e/pkg/skaffold/test/custom/custom.go#L83

Here is another example of specialcasing:
https://github.com/GoogleContainerTools/skaffold/blob/ba2390db18ad15ccf1eefbce39d6dd8f34016914/pkg/skaffold/build/custom/script_test.go#L141

Should I disable it on windows like this?:
https://github.com/GoogleContainerTools/skaffold/blob/ba2390db18ad15ccf1eefbce39d6dd8f34016914/pkg/skaffold/docker/auth_test.go#L54

tejal29 reviewed (commented):
something went wrong with github and not all comment went through last time.

briandealwis commented on integration/examples/custom-tests/skaffold.yaml:
```diff
@@ -0,0 +1,17 @@
+        timeoutSeconds: 60
+        dependencies:
+          paths:
+          -  "*_test.go"
```
Add `test.sh` too!

PriyaModali replied:
Added

briandealwis replied:
I'm not seeing it!

briandealwis commented on pkg/skaffold/test/custom/custom.go:
```diff
@@ -0,0 +1,144 @@
+}
+
+// TestDependencies returns dependencies listed for a custom test
+func (ct *Runner) TestDependencies() ([]string, error) {
```
This should receive a `context.Context`, just like in `Test()` above.  Then you can replace `context.Background()` below.

PriyaModali replied:
Custom method implements this interface
https://github.com/GoogleContainerTools/skaffold/blob/31cf2f13ae2419c558ec12c6ccf60567f388e2a2/pkg/skaffold/test/types.go#L33
And had to follow the method signature ` TestDependencies() ([]string, error)`.

Is there a different way to send it 'context.Context`?

briandealwis commented on integration/examples/custom-tests/skaffold.yaml:
```diff
@@ -0,0 +1,17 @@
+  artifacts:
+  - image: skaffold-example
+test:
+  - image: skaffold-example
```
It would be useful to have another test here to exercise the `echo [a,b,c]` logic for both *nix and Windows.

PriyaModali replied:
Added another custom command instance with `echo ...`.

briandealwis reviewed (commented):
This is almost done.  The integration test needs another instance to verify the test-dependencies-from-a-command, particularly for Windows.

briandealwis commented on integration/examples/custom-tests/skaffold.yaml:
```diff
@@ -0,0 +1,18 @@
+        dependencies:
+          paths:
+          -  "*_test.go"
+      - command: echo Hello world!!
```
Could you add:
```
dependencies:
  command: echo "[main_test.go]"
```
Then we can test both styles of specifying test dependencies.

PriyaModali replied:
Added the suggested dependency.

briandealwis reviewed (approved):
Two small changes and good to go.

Note that the `./test.sh` will fail on Windows, but we're not running our integration tests on Windows.  I suspect we'll have many other fixes requires for integration tests on Windows.
